frontend/openchat-agent/src/utils/cachedEventsIterator.ts

```diff
@@ -0,0 +1,140 @@
+    db: IDBPDatabase<ChatSchema>,
+    eventIndexRange: IndexRange,
+    context: MessageContext,
+    startIndex: number,
+    ascending: boolean,
+    maxEvents: number,
+    maxMessages: number,
+    maxMissing: number,
+): Promise<[EventWrapper<ChatEvent>[], Set<number>]> {
+    const bound = ascending ? eventIndexRange[1] : eventIndexRange[0];
+    const iterator = await EventsIterator.create(db, context, startIndex, ascending, bound);
+
+    const events: EnhancedWrapper<ChatEvent>[] = [];
+    const missing = new Set<number>();
+    let messageCount = 0;
+    let expectedNextIndex: number = startIndex;
+    while (events.length < maxEvents) {
+        const next = await iterator.getNext();
+        if (next === undefined) {
+            let remainingMissingCount = Math.min(
+                maxMessages - messageCount,
+                maxEvents - events.length,
+            );
+            if (ascending) {
+                for (let i = expectedNextIndex; i <= bound; i++) {
+                    missing.add(i);
+                    if (--remainingMissingCount === 0) break;
+                }
+            } else {
+                for (let i = expectedNextIndex; i >= bound; i--) {
+                    missing.add(i);
+                    if (--remainingMissingCount === 0) break;
+                }
+            }
+            break;
+        }
+
+        events.push(next);
+
+        if (ascending) {
+            for (let i = expectedNextIndex; i < next.index; i++) {
+                missing.add(i);
+                if (missing.size > maxMissing) {
+                    break;
+                }
+            }
+
+            expectedNextIndex = next.index + 1;
+            if (expectedNextIndex > bound) break;
+        } else {
+            for (let i = expectedNextIndex; i > next.index; i--) {
+                missing.add(i);
+                if (missing.size > maxMissing) {
+                    break;
+                }
+            }
+
+            expectedNextIndex = next.index - 1;
+            if (expectedNextIndex < bound) break;
+        }
+
+        if (next.event.kind === "message") {
+            if (++messageCount == maxMessages) {
+                break;
+            }
+        }
+    }
+
+    return [events, missing];
+}
+
+class EventsIterator {
+    private hasStarted: boolean = false;
+
+    private constructor(
+        private cursor?: IDBPCursorWithValue<
+            ChatSchema,
+            ("chat_events" | "thread_events")[],
+            "chat_events" | "thread_events"
+        > | null,
+        private onComplete?: () => Promise<void>,
+    ) {}
+
+    static async create(
+        db: IDBPDatabase<ChatSchema>,
+        messageContext: MessageContext,
+        startIndex: number,
+        ascending: boolean,
+        bound: number,
+    ): Promise<EventsIterator> {
+        if ((ascending && startIndex > bound) || (!ascending && startIndex < bound)) {
+            return new EventsIterator();
```

Comment:
Is this not an error scenario? Should we be throwing?

Reply:
Yeah it means the args are invalid. I'll make it throw